include/forward_index_builder.hpp

```diff
@@ -0,0 +1,288 @@
+
+    void build(std::istream &             is,
+               std::string const &        output_file,
+               read_record_function_type  next_record,
+               process_term_function_type process_term,
+               std::ptrdiff_t             batch_size,
+               std::size_t                threads)
+    {
+        Document_Id    first_document{0};
+        std::ptrdiff_t batch_number = 0;
+
+        std::vector<Record> record_batch;
+        tbb::task_group     batch_group;
+        tbb::concurrent_bounded_queue<int> queue;
+        queue.set_capacity((threads - 1) * 2);
+        while (true) {
+            std::optional<Record> record = std::nullopt;
+            try {
+                if (not (record = next_record(is))) {
+                    auto last_batch_size = record_batch.size();
+                    Batch_Process bp{
+                        batch_number, std::move(record_batch), first_document, output_file};
+                    queue.push(0);
+                    batch_group.run([bp = std::move(bp), process_term, this, &queue]() {
+                        run(std::move(bp), process_term);
+                        int x;
+                        queue.try_pop(x);
+                    });
+                    ++batch_number;
+                    first_document += last_batch_size;
+                    break;
+                }
+            } catch (Warc_Format_Error &err) {
+                continue;
+            }
+            if (record->type() != "response") {
+                continue;
+            }
+            record_batch.push_back(std::move(*record)); // AppleClang is missing value() in Optional
+            if (record_batch.size() == batch_size) {
+                Batch_Process bp{
+                    batch_number, std::move(record_batch), first_document, output_file};
+                queue.push(0);
+                batch_group.run([bp = std::move(bp), process_term, this, &queue]() {
+                    run(std::move(bp), process_term);
+                    int x;
+                    queue.try_pop(x);
+                });
+                ++batch_number;
+                first_document += batch_size;
+                record_batch = std::vector<Record>();
+            }
+        }
+        batch_group.wait();
+        merge(output_file, static_cast<ptrdiff_t>(first_document), batch_number);
+    }
+};
+
+class Plaintext_Record {
+   public:
+    Plaintext_Record() = default;
+    Plaintext_Record(std::string trecid, std::string content)
+        : m_trecid(std::move(trecid)), m_content(std::move(content)) {}
+    [[nodiscard]] auto content() -> std::string & { return m_content; }
+    [[nodiscard]] auto content() const -> std::string const & { return m_content; }
+    [[nodiscard]] auto trecid() -> std::string & { return m_trecid; }
+    [[nodiscard]] auto trecid() const -> std::string const & { return m_trecid; }
+    [[nodiscard]] auto type() const noexcept -> std::string { return "response"; }
```

Comment:
I would remove this

Reply:
Yeah, if we write a `valid` function, this won't be needed.